packages/governance/src/votes/votes.cairo

```diff
@@ -0,0 +1,312 @@
+/// governance decisions. Voting power must be delegated to be counted, and an account can
+/// delegate to itself if it wishes to vote directly.
+///
+/// This component offers a unified interface for voting mechanisms across ERC20 and ERC721
+/// token standards, with the potential to be extended to other token standards in the future.
+/// It's important to note that only one token implementation (either ERC20 or ERC721) should
+/// be used at a time to ensure consistent voting power calculations.
+#[starknet::component]
+pub mod VotesComponent {
+    // We should not use Checkpoints or StorageArray as they are for ERC721Vote
+    // Instead we can rely on Vec
+    use core::num::traits::Zero;
+    use openzeppelin_account::dual_account::{DualCaseAccount, DualCaseAccountTrait};
+    use openzeppelin_governance::votes::interface::{IVotes, TokenVotesTrait};
+    use openzeppelin_governance::votes::utils::{Delegation};
+    use openzeppelin_introspection::src5::SRC5Component;
+    use openzeppelin_token::erc20::ERC20Component;
+    use openzeppelin_token::erc20::interface::IERC20;
+    use openzeppelin_token::erc721::ERC721Component;
+    use openzeppelin_token::erc721::interface::IERC721;
+    use openzeppelin_utils::nonces::NoncesComponent::InternalTrait as NoncesInternalTrait;
+    use openzeppelin_utils::nonces::NoncesComponent;
+    use openzeppelin_utils::structs::checkpoint::{Checkpoint, Trace, TraceTrait};
+    use starknet::ContractAddress;
+    use starknet::storage::Map;
+    use super::{OffchainMessageHash, SNIP12Metadata};
+
+    #[storage]
+    struct Storage {
+        Votes_delegatee: Map::<ContractAddress, ContractAddress>,
+        Votes_delegate_checkpoints: Map::<ContractAddress, Trace>,
+        Votes_total_checkpoints: Trace,
+    }
+
+    #[event]
+    #[derive(Drop, PartialEq, starknet::Event)]
+    pub enum Event {
+        DelegateChanged: DelegateChanged,
+        DelegateVotesChanged: DelegateVotesChanged,
+    }
+
+    #[derive(Drop, PartialEq, starknet::Event)]
+    pub struct DelegateChanged {
+        #[key]
+        pub delegator: ContractAddress,
+        #[key]
+        pub from_delegate: ContractAddress,
+        #[key]
+        pub to_delegate: ContractAddress
+    }
+
+    #[derive(Drop, PartialEq, starknet::Event)]
+    pub struct DelegateVotesChanged {
+        #[key]
+        pub delegate: ContractAddress,
+        pub previous_votes: u256,
+        pub new_votes: u256
+    }
+
+    pub mod Errors {
+        pub const FUTURE_LOOKUP: felt252 = 'Votes: future Lookup';
+        pub const EXPIRED_SIGNATURE: felt252 = 'Votes: expired signature';
+        pub const INVALID_SIGNATURE: felt252 = 'Votes: invalid signature';
+    }
+
+    #[embeddable_as(VotesImpl)]
+    impl Votes<
+        TContractState,
+        +HasComponent<TContractState>,
+        impl Nonces: NoncesComponent::HasComponent<TContractState>,
+        impl TokenTrait: TokenVotesTrait<ComponentState<TContractState>>,
```

Comment:
```suggestion
        +TokenVotesTrait<ComponentState<TContractState>>,
```
We don't need to name the implementation since we are not directly using it here.

Reply:
Good catch!